### DISCUSSION TITLE
(mini.icons): How to get the whole spec of icons for a particular category.

### DISCUSSION THREAD

**Discussion by voidstar-null** [nvim-mini/mini.nvim#1415] [Q&A] [answered]
As the title suggests, is there any way to fetch a table consisting of `name -> icon` mappings for a category?

**Comment by echasnovski** [comment, marked as answer]
The answer is both "yes" and "no".

The "yes" part is that it is indeed possible with the following:

```lua
_G.get_all_icons = function(category)
  local res = {}
  for _, name in ipairs(MiniIcons.list(category)) do
    local icon, hl = MiniIcons.get(category, name)
    res[name] = { glyph = icon, hl = hl }
  end
  return res
end
```

Then calling `_G.get_all_icons('filetype')` will get a map from `name` to icon data (glyph and highlight group).

The "no" part is that it will work for all but "extension" and "file" categories, because they [rely on `vim.filetype.match()`](https://github.com/echasnovski/mini.nvim/blob/4228f166ee9db3e910eea1915e2d3683028add47/doc/mini-icons.txt#L356-L357) to first get corresponding filetype and then get icon data from it. For example `_G.get_all_icons('extension').lua` will return `nil`, although `MiniIcons.get('extension', 'lua')` will return correct data.

**Reply by voidstar-null** [reply]
> _G.get_all_icons = function(category)
  local res = {}
  for _, name in ipairs(MiniIcons.list(category)) do
    local icon, hl = MiniIcons.get(category, name)
    res[name] = { glyph = icon, hl = hl }
  end
  return res
end

Thanks this is what I needed:-),  but wouldn't it be good if there exists a method within the plugin itself that does all this without having to loop through a whole list?
On reading the source code a bit,  I ended up with something like this:
```lua
MiniIcons.get_icons = function(category)
    return H[category .. '_icons']
end
```
As the purpose of this method is just to fetch the icon spec. but we cannot `H` outside of the 'icons.lua' file.


> The "no" part is that it will work for all but "extension" and "file" categories, because they [rely on vim.filetype.match()](https://github.com/echasnovski/mini.nvim/blob/4228f166ee9db3e910eea1915e2d3683028add47/doc/mini-icons.txt#L356-L357) to first get corresponding filetype and then get icon data from it. For example _G.get_all_icons('extension').lua will return nil, although MiniIcons.get('extension', 'lua') will return correct data.

Hmm, I tried doing it and it indeed does not work. May you explain , why? does that have something to do with inner workings of lua?

**Reply by echasnovski** [reply]
> but wouldn't it be good if there exists a method within the plugin itself that does all this without having to loop through a whole list?

I think exporting the minimum required to achieve any result is a good idea. Having function that returns *both* icon names and their data leads to duplicating responsibilities.

> On reading the source code a bit, I ended up with something like this:
> 
> ```lua
> MiniIcons.get_icons = function(category)
>     return H[category .. '_icons']
> end
> ```

Please don't do this. This is not safe, prone to breaking, and (most importantly) [doesn't 100% work](https://github.com/echasnovski/mini.nvim/blob/4228f166ee9db3e910eea1915e2d3683028add47/lua/mini/icons.lua#L877).

> Hmm, I tried doing it and it indeed does not work. May you explain , why? does that have something to do with inner workings of lua?

It has to do with not wanting to track all complicated rules about what different files and extensions actually represent. Relying on `vim.filetype.match()` allows to leverage a massive amount of work done in both Vim and Neovim by tracking these rules and having them up to date. Focusing only on support of all possibly reasonable filetypes leads to better outcome with less work.

**Reply by voidstar-null** [reply]
Got it! thanks again for the explanation :-)
